scene/main/window.cpp

```diff
@@ -1290,34 +1269,47 @@ void Window::set_theme(const Ref<Theme> &p_theme) {
 		return;
 	}
 
+	if (theme.is_valid()) {
+		theme->disconnect("changed", callable_mp(this, &Window::_theme_changed));
+	}
+
 	theme = p_theme;
+	if (theme.is_valid()) {
+		Control::_propagate_theme_changed(this, nullptr, this, is_inside_tree(), true);
+		theme->connect("changed", callable_mp(this, &Window::_theme_changed), CONNECT_DEFERRED);
+		return;
+	}
 
-	if (!p_theme.is_null()) {
-		theme_owner = nullptr;
-		theme_owner_window = this;
-		Control::_propagate_theme_changed(this, nullptr, this);
-	} else {
-		Control *parent_c = cast_to<Control>(get_parent());
-		if (parent_c && (parent_c->data.theme_owner || parent_c->data.theme_owner_window)) {
-			Control::_propagate_theme_changed(this, parent_c->data.theme_owner, parent_c->data.theme_owner_window);
-		} else {
-			Window *parent_w = cast_to<Window>(get_parent());
-			if (parent_w && (parent_w->theme_owner || parent_w->theme_owner_window)) {
-				Control::_propagate_theme_changed(this, parent_w->theme_owner, parent_w->theme_owner_window);
-			} else {
-				Control::_propagate_theme_changed(this, nullptr, nullptr);
-			}
-		}
+	Control *parent_c = Object::cast_to<Control>(get_parent());
+	if (parent_c && (parent_c->data.theme_owner || parent_c->data.theme_owner_window)) {
+		Control::_propagate_theme_changed(this, parent_c->data.theme_owner, parent_c->data.theme_owner_window, is_inside_tree(), true);
+		return;
 	}
+
+	Window *parent_w = cast_to<Window>(get_parent());
+	if (parent_w && (parent_w->theme_owner || parent_w->theme_owner_window)) {
+		Control::_propagate_theme_changed(this, parent_w->theme_owner, parent_w->theme_owner_window, is_inside_tree(), true);
+		return;
+	}
+
+	Control::_propagate_theme_changed(this, nullptr, nullptr, is_inside_tree(), true);
```

Comment:
This seems to ignore the case where the window has no parent (the root window).

Reply:
What do you mean? If the window has no parent then `parent_c` and `parent_w` will both be `nullptr`, and `Control::_propagate_theme_changed(this, nullptr, nullptr, is_inside_tree(), true);` will be called.

Reply:
I'm sorry, you're right, apologies for the reckless remarks earlier.